### PR TITLE
Benchmark macro

### LINKED ISSUE #4861: Create macros to simplify benchmarking.
The benchmarking setup introduced by https://github.com/paritytech/substrate/pull/4818 has a lot of boilerplate code which can be improved through the use of macros.

## `decl_module`

The `decl_module` macro should introduce a `#[benchmark(...)]` tag which can be placed over extrinsics. This can include a list of letters [A-Z] which represents parameters to test, and a range over which we should test them. Example:

```rust
#[benchmark([A, 1, 10], [B, 5, 50])]
fn transfer(origin, who: T::AccountId, amount: T::Balance) {
   ...
}
```

All extrinsics which have this attribute will:
* Be added to an Enum which will implement `BenchmarkingSetup` so that we can select which benchmark to run through the CLI.
* Have a lookup function which maps the string name of the function to its appropriate BenchmarkingSetup enum.

Adding the `#[benchmark(...)]` tag should also create an object that can be exposed into the `construct_runtime!` macro, so that we can know this pallet should expose it's benchmarking setup.

## `construct_runtime!`

The `construct_runtime!` macro should then implement the `Benchmark` runtime api, creating the `dispatch_benchmark` function and populating it with Pallets that have exposed benchmarking.

It would simply create a function which maps the appropriate strings to the correct `run_benchmark` function.

```rust
match module.as_slice() {
	b"pallet-identity" | b"identity" => Identity::run_benchmark(extrinsic, steps, repeat).ok(),
	_ => return None,
}

### PR DESCRIPTION
**Pull request by gavofyork**
Fixes #4861

### CODE REVIEW

**Comment by kianenigma** [comment]
Will it still be possible, given this, to manually build the setup? There is a situation in https://github.com/paritytech/substrate/pull/4947 that you can see and it cannot be really tracked with this benchmarking setup. One of the parameters is 

```rust
Enum Mode {
   BenchMode1,
   BenchMode2, 
}
```

and stepping over this is not possible.

**Comment by marcio-diaz** [comment]
> Will it still be possible, given this, to manually build the setup? There is a situation in #4947 that you can see and it cannot be really tracked with this benchmarking setup. One of the parameters is
> 
> ```rust
> Enum Mode {
>    BenchMode1,
>    BenchMode2, 
> }
> ```
> 
> and stepping over this is not possible.


> Will it still be possible, given this, to manually build the setup? There is a situation in #4947 that you can see and it cannot be really tracked with this benchmarking setup. One of the parameters is
> 
> ```rust
> Enum Mode {
>    BenchMode1,
>    BenchMode2, 
> }
> ```
> 
> and stepping over this is not possible.

I think you can make two benchmarks out of the enum and refactor the common parts.

File: frame/benchmarking/src/lib.rs [hunk 1]
```diff
@@ -14,79 +14,397 @@
...
+					),*
+				]
+			}
 
-	/// Set up the storage, and prepare a call and caller to test in a single run of the benchmark.
-	fn instance(&self, components: &[(BenchmarkParameter, u32)]) -> Result<(Call, RawOrigin), &'static str>;
+			fn instance(&self, components: &[($crate::BenchmarkParameter, u32)])
+				-> Result<(crate::Call<T>, RawOrigin<T::AccountId>), &'static str>
+			{
+				let _benchmarks_seed = 0;
```
Comment (marcio-diaz) [inline frame/benchmarking/src/lib.rs]:
```suggestion
```
`_benchmarks_seed` didn't work for me when trying to use it in `account()`.

**Comment by gnunicorn** [comment]
So, the reports say that it consistently slightly improves on wasm but is sometimes a regression on native ... how do you want to proceed?

**Comment by marcio-diaz** [comment]
> So, the reports say that it consistently slightly improves on wasm but is sometimes a regression on native ... how do you want to proceed?

Wrong PR :)? I'll merge once CI finish, other benchmarks depend on this.

**Pull request merged**